source/retryable-writes/tests/README.rst

```diff
@@ -121,14 +121,16 @@ Each YAML file has the following keys:
 Use as Integration Tests
 ========================
 
-Running these as integration tests will require a running mongod server. Each of
-these tests is valid against a standalone mongod, a replica set, and a sharded
-system for server version 3.6.0 or later.
+Running these as integration tests will require a running MongoDB cluster with
+server versions 3.6.0 or later. The ``{setFeatureCompatibilityVersion: 3.6}``
+admin command will also need to have been executed to enable support for
+retryable writes on the cluster.
 
 Network Error Tests
 ===================
 
-The YAML tests exercise the following test scenarios:
+The YAML tests may be used to test a replica set or shard cluster, whose shards
+are replica sets. The tests exercise the following scenarios:
```

Comment:
The tests can't run against a sharded cluster because the failpoint does not exist on a mongos.

Reply:
I hadn't considered that, but it makes sense as I've always run fail points on individual mongod nodes.

It wouldn't make sense to setup the fail point on the primaries within the shard cluster, as that would only assert mongos' correctness to retry writes within the cluster. If mongos doesn't provide us with a way to drop its connections (that would probably require its own fail point, assuming mongos even supports fail points), I suppose we can't test it at all.

Reply:
Bumping this comment thread, so we don't lose it due to diff rendering.

Reply:
> Perhaps as a prose test suggesting that folks use mongo-orchestration or some other tool to drop mongos nodes on demand?

I don't have any good ideas for this. Perhaps it's enough to test retries on a replica set and separately test that lsid/txnNumber are sent to mongos?

Reply:
I like this suggestion.  We prove driver and server work correctly together by testing against a repl set with failpoints.  We prove the driver works properly against a shard cluster with command monitoring-based tests that assert presence of the lsid/txnNumber.

Reply:
I added a "Command Construction Tests" section at the bottom, which talks about using APM to assert that write commands include/exclude a transaction ID based on whether the operation is supported or unsupported, respectively.